spec.bs

```diff
@@ -6,57 +6,940 @@ Status: CG-DRAFT
+    1. Let |epochLength| be [=budget lifetime=].
+    1. Let |unsafeMoment| be the [=wall clock=]'s [=wall clock/unsafe current time=].
+    1. Let |moment| be the result of running [=coarsen time=] algorithm given |unsafeMoment| and [=wall clock=] as input.
+    1. Let |fromUnixEpochTime| be the [=duration from=] the [=Unix epoch=] to |moment|.
+    1. Let |threshold| be |fromUnixEpochTime| minus |epochLength|.
+    1. Let |debitSum| be 0.
+    1. For each [=list/item=] |bitDebit| in |ledger|, do the following steps:
+        1. Let |timestamp| be |bitDebit|'s [=bit debit/timestamp=].
+        1. Let |debit| be |bitDebit|'s [=bit debit/bits=].
+        1. If |timestamp| is greater than or equal to |threshold|, then increment |debitSum| by |debit|.
+    1. Let |remainingBudget| be |maxBits| minus |debitSum|.
+    1. Return |remainingBudget|.
+  </div>
 
-dictionary SharedStorageUrlWithMetadata {
-  required USVString url;
-  object reportingMetadata;
-};
+  A [=bit debit=] will need to be charged to the [=shared storage budget table=] for each top-frame navigation initiated by a fenced frame whose source was generated via {{WindowSharedStorage/selectURL()}}, as this can leak cross-site data. Since the [=bit debit/bits=] to charge is calculated during the call to {{WindowSharedStorage/selectURL()}} but only actually recorded in the [=shared storage budget table=] if and when the resulting fenced frame initiates a top-frame navigation, the [=bit debit/bits=] must be stored as a <dfn>pending shared storage budget debit</dfn> in the corresponding {{FencedFrameConfig}} until this time.
+
+  Between [=beginning navigation=] and [=ending navigation=], a [=user agent=] will perform the [=charge shared storage budget=] algorithm.
+
+  <div algorithm>
+    To <dfn>charge shared storage budget</dfn> during a [=beginning navigation|navigation=] with [=navigable=] |navigable| and {{Document}} |sourceDocument|, run the following steps:
+
+    1. If |navigable| is not a [=navigable/traversable navigable=], return.
+    1. Let |node| be |sourceDocument|'s [=node navigable=].
+    1. While |node| is not null:
+        1. If |node| has a {{FencedFrameConfig}}:
+            1. Let |origin| be |node|'s [=active document=]'s [=document/origin=].
+            1. Let |pendingBits| be |node|'s {{FencedFrameConfig}}'s [=pending shared storage budget debit=].
+            1. If |pendingBits| is greater than 0, run the following steps:
+                1. Let |ledger| be [=user agent=]'s [=shared storage budget table=][|origin|].
+                1. Let |bitDebit| be a new [=bit debit=].
+                1. Set |bitDebit|'s [=bit debit/bits=] to |pendingBits|.
+                1. Let |unsafeMoment| be the [=wall clock=]'s [=wall clock/unsafe current time=].
+                1. Let |moment| be the result of running [=coarsen time=] algorithm given |unsafeMoment| and [=wall clock=] as input.
```

Comment:
Again, `[=current wall time=]` is the coarsened wall clock's current time.  So I think you can simply those two steps into one.